Add Cloud Run (Golang) Hello World

grayside opened the pull request:
This is an initial port of cloud-run-hello for use as a sample.

I've annotated it with a few TODOs/considerations for us to work out further definition of what these should look like.

All SVGs were copied verbatim from the cloud-run-hello sample.

* [ ] Golang Code Review
* [ ] Finalize sample design
* [ ] Complete Cloud Code integration
* [ ] Complete README
* [ ] Switch integration tests to Go instead of BASH

grayside commented on cloudbuild.meta.yaml:
```diff
@@ -46,6 +46,7 @@ steps:
     declare -a apps=("hello-world" "guestbook")
     gcloud builds submit . --config cloudbuild.tmpl.yaml --substitutions _DIR=${dir},_LANG=${lang},_APP=${apps[0]}
     gcloud builds submit . --config cloudbuild.tmpl.yaml --substitutions _DIR=${dir},_LANG=${lang},_APP=${apps[1]}
+    gcloud builds submit . --config golang/go-cloud-run-hello-world
```
Per discussion with Simon, we do not need to use the substitutions since we're using a unique cloudbuild.yaml per sample.

grayside commented on golang/go-cloud-run-hello-world/.gitignore:
```diff
@@ -0,0 +1,14 @@
+# Binaries for programs and plugins
```
This file was copied from the `go-hello-world` sample.

grayside commented on golang/go-cloud-run-hello-world/.idea/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
This file was copied from the go-hello-world sample.

grayside commented on golang/go-cloud-run-hello-world/.idea/externalDependencies.xml:
```diff
@@ -0,0 +1,6 @@
+    <component name="ExternalDependencies">
+          <plugin id="com.google.gct.core" />
+    </component>
+</project>
```
This file was copied from the `go-hello-world` sample.

grayside commented on golang/go-cloud-run-hello-world/.idea/go-cloud-run-hello-world.iml:
```diff
@@ -0,0 +1,8 @@
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
This file was copied from the `go-hello-world` sample.

grayside commented on golang/go-cloud-run-hello-world/.idea/modules.xml:
```diff
@@ -0,0 +1,8 @@
+      <module fileurl="file://$PROJECT_DIR$/.idea/go-cloud-run-hello-world.iml" filepath="$PROJECT_DIR$/.idea/go-cloud-run-hello-world.iml" />
+    </modules>
+  </component>
+</project>
```
This file was copied from the `go-hello-world` sample and the module reference updated to the dirname of this sample.

grayside commented on golang/go-cloud-run-hello-world/.vscode/launch.json:
```diff
@@ -0,0 +1,21 @@
+      "program": "${workspaceFolder}/."
+    },
+  ]
+}
```
I'm not sure what to do with this file, especially given the discussion to leave skaffold out of the Cloud Run samples.

grayside commented on golang/go-cloud-run-hello-world/.vscode/tasks.json:
```diff
@@ -0,0 +1,42 @@
+            "problemMatcher": [],
+        }
+    ]
+}
```
I'm not sure what to do with this file, especially given the discussion to leave skaffold out of the Cloud Run samples.

grayside commented on golang/go-cloud-run-hello-world/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+# If you want to run WITHOUT the debugging server:
+#   * COMMENT OUT the previous ENTRYPOINT statements,
+#   * UNCOMMENT the following ENTRYPOINT statement.
+# ENTRYPOINT ["/app"]
```
This file was copied from `go-hello-world`.

Changes include:
* adding static assets
* swapping in the -mod=readonly directive used in the Cloud Run samples.

Normally for golang I'd use a multi-stage Dockerfile, but not sure what the right way is to split the difference with local debugging, so I'm following the existing sample. If we're using this file as a model for a production Dockerfile, it might be worth doing something more complicated.

grayside commented on golang/go-cloud-run-hello-world/README.md:
````diff
@@ -0,0 +1,53 @@
+
+```
+gcloud builds submit . --config cloudbuild.yaml --substitutions COMMIT_SHA=manual
+```
````
Started sketching out this README, but it seems like there may be some flux on what we want to include.

grayside commented on golang/go-cloud-run-hello-world/cloudbuild.yaml:
```diff
@@ -0,0 +1,67 @@
+          --platform managed --region ${_CLOUDSDK_REGION}
+    }
+    echo $(get_url) > _service_url
+    echo "Cloud Run URL for ${_SERVICE} is $(cat _service_url)"
```
This was adapted from the `go-hello-world` sample. Since the `test_content.sh` is no longer trying to be general-purpose this could be moved into it to simplify the `cloudbuild.yaml`.

grayside commented on golang/go-cloud-run-hello-world/cloudbuild.yaml:
```diff
@@ -0,0 +1,67 @@
+  - '${_SERVICE}'
+  - '--image'
+  - 'gcr.io/${PROJECT_ID}/${_SERVICE_TAG}:${COMMIT_SHA}'
+  - '--allow-unauthenticated'
```
We should probably remove `--allow-unauthenticated` and instead run the integration tests with an identity token via `gcloud auth print-identity-token`. Kept this here for initial simplicity.

grayside commented on golang/go-cloud-run-hello-world/cloudbuild.yaml:
```diff
@@ -0,0 +1,67 @@
+  _CLOUDSDK_VERSION: latest
+  _CLOUDSDK_REGION: us-central1
+  _SERVICE: hello-world
+  _SERVICE_TAG: go-cloud-run-hello-world
```
This sample name is really long so I didn't want to use it as the service URL itself if developers might manually type it.

grayside commented on golang/go-cloud-run-hello-world/index.html:
```diff
@@ -0,0 +1,207 @@
+<!doctype html>
```
This file was copied as-is from cloud-run-hello. There are open issues in the repo to tune it for accessibility so we'll want to track and port changes

grayside replied:
Also, this page has gotten complicated as it emphasizes a strong out-of-box experience as a running app, rather than being a sample HTML page. Do we want to maintain this as-is in multiple languages or simplify?

simonz130 replied:
When you say simplify - do you mean having the same template HTML referenced in multiple languages templates?

grayside replied:
The current webpage involves multiple images, complicated CSS, support for dark mode, etc. It is complicated enough that it's had a number of changes recently to improve accessibility, performance, and so on: in other words, it's not really maintained as a sample but as an experience.

Do we want to track the cloud-run-hello or fork? If we want to fork, we should consider simplifying. For example, instead of the full web page currently generated, we could go with:

text/plain:
```
Service ??? running Revision ??? in Project ???
```

I asked this question because I don't have a strong sense of the value of the more visual experience in this case.

simonz130 replied:
The value of more visual experience is giving developers better sense of accomplishment. It also demoes better compared to a plain text.
I agree, we should fork and simplify where appropriate, we don't need a dark mode or other more advanced UI-related scenarios or features.

grayside commented on golang/go-cloud-run-hello-world/main.go:
```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Google LLC
```
I updated this sample from the original to better align with Cloud Run sample practices, including more error handling and use of the metadata client library.

grayside commented on golang/go-cloud-run-hello-world/test_content.sh:
```diff
@@ -0,0 +1,73 @@
+if [[ -z "$url" ]]; then
+  echo "Failed: URL not specified."
+  exit 1
+fi
```
I added this stanza because during development of the `cloudbuild.yaml` I had an error where the URL was an empty value and the test process went through all retries for no reason.

simonz130 commented on golang/go-cloud-run-hello-world/README.md:
```diff
@@ -0,0 +1,53 @@
+
+1. [Getting Started](#getting-started])
+2. [What's in the box](https://cloud.google.com/code/docs/vscode/quickstart#whats_in_the_box)
+3. [Using the Command Line](#using-the-command-line)
```
We shouldn't have the section of using command line (even though it should work when invoking appropriate command line). This is because we optimize for Cloud Code IDE experience (IntelliJ and vscode).

simonz130 replied:
We probably need two more sections:

1) Running locally
2) Running on Cloud Run

simonz130 replied:
Also, maybe "Next steps" with more useful links.

grayside replied:
Alright, will iterate this. However, I don't think it's good to hide the command-line operations. It might make sense for a simple example like this, but once it's more complicated understanding the escape valve is important.

grayside replied:
Mostly done. I don't think all the useful links exist yet.

simonz130 commented on golang/go-cloud-run-hello-world/README.md:
```diff
@@ -0,0 +1,53 @@
+----
+## Table of Contents
+
+### Cloud Code for Visual Studio Code
```
Our instructions should equally target VSCode and IntellIJ. Let's just remove this header and have "Table of contents" only.

grayside replied:
This is copied from other sample READMEs, I'll iterate this for Cloud Run, but maybe there are changes to make to the existing sample READMEs?

grayside replied:
Made some structural changes to try covering both IDEs more even-handedly. There's gaps for IntelliJ still.

simonz130 commented on golang/go-cloud-run-hello-world/README.md:
````diff
@@ -0,0 +1,53 @@
+## Build the Container Image
+
+```
+docker build -t go-cloud-run-hello-world .
````
This should be transparent for users of Cloud Code: it will build container image as part of trying to deploy the app locally or to Cloud Run.

grayside replied:
Removed this given transparency and determination not to include CLI steps.

simonz130 commented on golang/go-cloud-run-hello-world/README.md:
````diff
@@ -0,0 +1,53 @@
+## Run the Container Locally
+
+```
+docker run ...
````
Likely will turn into:

* VSCode: run **Cloud Code: Run Locally** command on command pallete.
* IntelliJ: something along similar lines

simonz130 commented on golang/go-cloud-run-hello-world/README.md:
````diff
@@ -0,0 +1,53 @@
+docker run ...
+```
+
+## Deploy to Cloud Run
````
The actual command will be abstracted away by Cloud Code extension.

simonz130 commented on golang/go-cloud-run-hello-world/cloudbuild.yaml:
```diff
@@ -0,0 +1,67 @@
+  - '-c'
+  - |
+    get_url() {
+        gcloud run services describe ${_SERVICE} --format 'value(status.url)' \
```
We need to have isolation for these templates: in case multiple CI runs are happening at the same time, they shouldn't have conflicts w.r.t names. In kubernetes it's easy to achieve with namespaces. Not sure how to do it with Cloud Run.

grayside replied:
Good callout! The current best approach is to avoid duplication by naming the service distinctly, examples would be something like:

* hello-world-$SHORT_SHA
* hello-world-$BUILD_ID

If there can be a build that runs multiple samples, we can re-introduce the language:

* hello-world-go-$SHORT_SHA
* hello-world-go-$BUILD_ID
* hello-world-go-$_PR_NUMBER

PR Number would require some extra finesse if there's anything like a nightly build of all samples.

Happy to discuss future options but this is the best practice for now. Do you have a preference?

simonz130 replied:
<lang>-hello-world-cloud-run-$short_sha 
The same with $build_id sounds good too. Up to you.

I'm not familiar with Cloud Run networking, so what happens when you want to expose your Cloud Run service to internet? Will its IP object have a name that you can manage, like service/load balancer in kubernetes? Or is it internally managed by Cloud Run?
Asking to make sure we avoid conflicts in that too.

grayside replied:
No need to worry about ip addresses, that's completely abstracted by the platform.

grayside replied:
Ended up going with COMMIT_SHA to avoid multiple substitutions. Will switch the whole thing to SHORT_SHA if something proves too long.

simonz130 commented on golang/go-cloud-run-hello-world/main.go:
```diff
@@ -0,0 +1,101 @@
+	"net/http"
+	"os"
+
+	"cloud.google.com/go/compute/metadata"
```
How this will work when the template will be running locally?

grayside replied:
Current behavior is wait for timeout then it falls back to default `???`.
We could preempt the metadata server request by checking for a `GOOGLE_CLOUD_PROJECT` env var. Getting the Project ID on Cloud Run has a metadata server dependency.

simonz130 replied:
I like that!

Check the GOOGLE_CLOUD_PROJECT env var, if it's set, wait for timeout and fall back if needed. If it's not set, it means you run locally.

grayside replied:
Done.

simonz130 commented on golang/go-cloud-run-hello-world/main.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+func main() {
+	// Comment out in local development for timestamp prefixes.
```
Can this be implemented such that there will be no required manual step to get this to run locally vs on Cloud Run?

grayside replied:
This isn't required to run locally. It's just nice to have in terminal output, whereas it's a distraction in stackdriver.

Options:
1. Remove and always have timestamps
2. Leave in place and let developers edit the code if they want to.
3. Add a check for something like `DEV=1` env var and toggle accordingly.

The plan is for Cloud Code to simulate all the mechanisms we might use to detect running in Cloud Run.

simonz130 replied:
Could we generalize option 3 a little bit? So instread of introducing a new environment variable for local vs. cloud, we can test the same logic you already have for determining whether to call metadata server.

grayside replied:
Leaning on GOOGLE_CLOUD_PROJECT env var for this. It's a bit awkward for conceptual flow but adaptive is better than commenting instructions.

simonz130 commented on golang/go-cloud-run-hello-world/main.go:
```diff
@@ -0,0 +1,101 @@
+	// PORT environment variable is provided by Cloud Run.
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
```
Do we need to mention in the readme that it runs on 8080 when deployed locally?

grayside replied:
8080 is the default and the port is output as part of server start. That seems sufficient to me unless the goal is to document the availability of the PORT var to avoid port conflicts? If so, how does that relate to only using Cloud Code command palette to run locally?

simonz130 replied:
In case developers run multiple workloads in containers locally, we may want to let them know that this sample is using $PORT to expose the service locally.

grayside replied:
Added README section for `PORT` and `GOOGLE_CLOUD_PROJECT`

grayside commented on golang/go-cloud-run-hello-world/.vscode/launch.json:
```diff
@@ -0,0 +1,21 @@
+  "configurations": [
+    {
+      "name": "Run on Cloud Run",
+      "type": "cloudcode.kubernetes",
```
Done

grayside commented:
I've complete a feedback response round. Updates are mostly README revisions, otherwise there are several conversation threads exploring further changes/requirements.

grayside commented:
Completed last round on this PR for today. Thanks for taking the time to go around twice!

grayside commented:
Changes in recent updates:

* Updated Congrats => Congratulations
* Refactored integration test to a go implementation
* Added local service CLI testing instructions to README
